File: @@api/src/main/java/com/example/nb_weather_ol/converter/FlexibleLocalDateAttributeConverter.java
package com.example.nb_weather_ol.converter;

import jakarta.persistence.AttributeConverter;
import jakarta.persistence.Converter;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.Optional;

@Converter(autoApply = true)
public class FlexibleLocalDateAttributeConverter implements AttributeConverter<LocalDate, String> {

    private static final DateTimeFormatter[] FORMATTERS = {
            DateTimeFormatter.ofPattern("yyyy-MM-dd"),
            DateTimeFormatter.ofPattern("dd-MM-yyyy"),
            DateTimeFormatter.ofPattern("MM/dd/yyyy"),
            DateTimeFormatter.ofPattern("yyyy/MM/dd")
    };

    @Override
    public String convertToDatabaseColumn(LocalDate localDate) {
        return Optional.ofNullable(localDate)
                .map(date -> date.format(DateTimeFormatter.ISO_LOCAL_DATE))
                .orElse(null);
    }

    @Override
    public LocalDate convertToEntityAttribute(String dateString) {
        if (dateString == null || dateString.trim().isEmpty()) {
            return null;
        }

        for (DateTimeFormatter formatter : FORMATTERS) {
            try {
                return LocalDate.parse(dateString, formatter);
            } catch (Exception e) {
                // Try next formatter
            }
        }

        throw new IllegalArgumentException("Unable to parse date: " + dateString);
    }
}

File: backend-kotlin/src/main/kotlin/com/nbweatherproject/backend/converter/FlexibleLocalDateAttributeConverter.kt
package com.nbweatherproject.backend

import jakarta.persistence.AttributeConverter
import jakarta.persistence.Converter
import java.time.LocalDate
import java.time.format.DateTimeFormatter

@Converter(autoApply = true)
class FlexibleLocalDateAttributeConverter : AttributeConverter<LocalDate, String> {

    companion object {
        private val FORMATTERS = arrayOf(
            DateTimeFormatter.ofPattern("yyyy-MM-dd"),
            DateTimeFormatter.ofPattern("dd-MM-yyyy"),
            DateTimeFormatter.ofPattern("MM/dd/yyyy"),
            DateTimeFormatter.ofPattern("yyyy/MM/dd")
        )
    }

    override fun convertToDatabaseColumn(localDate: LocalDate?): String? {
        return localDate?.format(DateTimeFormatter.ISO_LOCAL_DATE)
    }

    override fun convertToEntityAttribute(dateString: String?): LocalDate? {
        if (dateString.isNullOrBlank()) {
            return null
        }

        for (formatter in FORMATTERS) {
            try {
                return LocalDate.parse(dateString, formatter)
            } catch (e: Exception) {
                // Try next formatter
            }
        }

        throw IllegalArgumentException("Unable to parse date: $dateString")
    }
}

